### ISSUE TITLE
On network failure and MQTT reconnection, multiple shadow updates are sent

### ISSUE THREAD

**Issue by psiphi75** [aws/aws-iot-device-sdk-cpp-v2#435]
### Describe the bug

This issue may lie in `aws-c-mqtt` or it's implementation of it, so I'm raising it here.

MQTT messages can get stuck in the not-acknowled state.  This causes some messages to be resent every time the network reconnects to the cloud.  Some of these messages can be days or weeks old.  This causes issues in the shadow with incorrect state.

### Expected Behavior

I'd expect all messages to be ACK'd eventually such that we only ever send one message.  I understand that setting QOS_1 we can expect more than one message sent on occasion, but this bug is beyond that.

### Current Behavior

I can send the full logs to an AWS employee, but I'm not posting them publicly.

Below are the mix of AWS logs and application logs.  It shows that message 12 is sent just after the network went down, no ACK was received.  The message was correctly resent, an ACK was received, but there was an error.

```
2022-05-23 01:39:02.892 ( 454.846s) [main thread     ]    NetlinkListener.cpp:124   INFO| Interface ppp0 is DOWN

[TRACE] [2022-05-23T01:39:08Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: Attempting send of publish 12 first attempt
[TRACE] [2022-05-23T01:39:08Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: request 12 sent, but waiting on an acknowledgement from peer.

2022-05-23 01:40:09.297 ( 521.252s) [main thread     ]    NetlinkListener.cpp:119   INFO| Interface ppp0 is UP

[TRACE] [2022-05-23T01:40:17Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: Attempting send of publish 12 resend
[TRACE] [2022-05-23T01:40:17Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: request 12 sent, but waiting on an acknowledgement from peer.
[TRACE] [2022-05-23T01:40:18Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: received ack for message id 12
[TRACE] [2022-05-23T01:40:18Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: message id 12 completed with error code 0, removing from outstanding requests list.
[DEBUG] [2022-05-23T01:40:18Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: received completion for message id 12 but no outstanding request exists.  Assuming this is an ack of a resend when the first request has already completed.

2022-05-23 01:44:29.817 ( 781.772s) [main thread     ]    NetlinkListener.cpp:124   INFO| Interface ppp0 is DOWN
2022-05-23 01:45:41.163 ( 853.117s) [main thread     ]    NetlinkListener.cpp:119   INFO| Interface ppp0 is UP

[TRACE] [2022-05-23T01:45:54Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: received ack for message id 12
[TRACE] [2022-05-23T01:45:54Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: message id 12 completed with error code 0, removing from outstanding requests list.
[DEBUG] [2022-05-23T01:45:54Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: received completion for message id 12 but no outstanding request exists.  Assuming this is an ack of a resend when the first request has already completed.

2022-05-23 01:46:45.967 ( 917.922s) [main thread     ]    NetlinkListener.cpp:124   INFO| Interface ppp0 is DOWN
2022-05-23 01:47:53.232 ( 985.186s) [main thread     ]    NetlinkListener.cpp:119   INFO| Interface ppp0 is UP

[TRACE] [2022-05-23T01:48:00Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: received ack for message id 12
[TRACE] [2022-05-23T01:48:00Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: message id 12 completed with error code 0, removing from outstanding requests list.
[DEBUG] [2022-05-23T01:48:00Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: received completion for message id 12 but no outstanding request exists.  Assuming this is an ack of a resend when the first request has already completed.

2022-05-23 01:50:26.311 (1138.266s) [main thread     ]    NetlinkListener.cpp:124   INFO| Interface ppp0 is DOWN
2022-05-23 01:50:26.333 (1138.288s) [main thread     ]    NetlinkListener.cpp:119   INFO| Interface ppp0 is UP

[TRACE] [2022-05-23T01:50:34Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: received ack for message id 12
[TRACE] [2022-05-23T01:50:34Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: message id 12 completed with error code 0, removing from outstanding requests list.
[DEBUG] [2022-05-23T01:50:34Z] [a49fe9f0] [mqtt-client] - id=0x99cd38: received completion for message id 12 but no outstanding request exists.  Assuming this is an ack of a resend when the first request has already completed.
```

### Reproduction Steps

The following steps are used to reproduce:

- Connect the shadow to AWS IoT
- Disconnect the network
- Update the shadow
- Reconnect the network
- Wait for shadow to reconnect (it get's published), but the ACK is not correctly handled.
- Disconnect the network.
- Reconnect the network
- Wait for the shadow to reconnect (old MQTT messages get published, **these messages can be days old**).

### Possible Solution

In `aws-c-mqtt` in the `mqtt_request_complete()` function correctly handle the removal of the message.  But I could be wrong.

### Additional Information/Context

When the network is lost our application will explicitly call `MqttConnection->Disconnect()` and `Connect()` for when the network is available again.

### SDK version used

1.17.0

### Environment details (OS name and version, etc.)

Linux 5.4.0, armv7

**Comment by jmklix** [comment]
I think you want to be using the shadow version to ensure that you only use the latest and most up to date value. As you noticed QOS_1 allows a message to be sent multiple times which looks like what is happening to you. You can use the [version number](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/ca6fd530e718e04fd2926575372f27a58079189c/shadow/include/aws/iotshadow/ShadowDeltaUpdatedEvent.h#L48) to know when to discard message deltas. This is not something that the shadow sample uses so that might be causing some confusion. Please let me know if that fixes your problem?

**Comment by psiphi75** [comment]
TLDR; this is a reproducible bug where old MQTT messages miss the MQTT ACK message, hence are re-published every time the AWS C++ SDK reconnects to AWS IoT.  I'll explain the details below.

> I think you want to be using the shadow version to ensure that you only use the latest and most up to date value.

I have tried this.  But the shadow in AWS IoT get incremented each time a shadow update occurs, even if an stale MQTT message arrives.  See the example shadow below.

> As you noticed QOS_1 allows a message to be sent multiple times which looks like what is happening to you.

I understand how QOS_1 is supposed to work.  But it's not working how it's suppose to work.  My worked example in the above bug report clearly explains how ACK messages from the AWS MQTT broker are not applied to the pending MQTT packet.

> You can use the [version number](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/ca6fd530e718e04fd2926575372f27a58079189c/shadow/include/aws/iotshadow/ShadowDeltaUpdatedEvent.h#L48) to know when to discard message deltas. This is not something that the shadow sample uses so that might be causing some confusion. 

This is not about shadow deltas, this is when the AWS IoT SDK publishes to the cloud.

> Please let me know if that fixes your problem?

No, I feel like my original bug report was not understood.  Let me know if you want me to explain it in more detail.

Below is the example shadow.  It shows the effect of a very stale MQTT message (a message that had an "MQTT ACK", but is still being sent) coming in and trumping the most recent shadow state.  You can see this message (id 9) from the trace logs, the 

```text
[TRACE] [2022-05-26T20:40:39Z] [a6f4c000] [mqtt-client] - id=0x879678: Currently not in the event-loop thread, scheduling a task to send message id 9.
[DEBUG] [2022-05-26T20:41:57Z] [a4afe9f0] [mqtt-client] - id=0x879678: received completion for message id 9 but no outstanding request exists.  Assuming this is an ack of a resend when the first request has already completed.
[DEBUG] [2022-05-26T20:46:09Z] [a4afe9f0] [mqtt-client] - id=0x879678: received completion for message id 9 but no outstanding request exists.  Assuming this is an ack of a resend when the first request has already completed.
[DEBUG] [2022-05-26T20:49:39Z] [a4afe9f0] [mqtt-client] - id=0x879678: received completion for message id 9 but no outstanding request exists.  Assuming this is an ack of a resend when the first request has already completed.
```

At `2022-05-26T20:40:39Z` the message was queued to be sent, it was sent at around `2022-05-26T20:41:57Z` and ACK received, but the ACK was not applied and the message was sent again many times.


```JSON
{
  "previous": {
    "state": {
      "reported": {
        "alarms": {
          "ALR008": {
            "alarmTriggered": false,
            "counter": 5,
            "triggeredAt": 1653598071
          }
        }
      }
    },
    "metadata": {
      "reported": {
        "alarms": {
          "ALR008": {
            "alarmTriggered": {
              "timestamp": 1653598074
            },
            "counter": {
              "timestamp": 1653598074
            },
            "triggeredAt": {
              "timestamp": 1653598074
            }
          }
        }
      }
    },
    "version": 540
  },
  "current": {
    "state": {
      "reported": {
        "alarms": {
          "ALR008": {
            "alarmTriggered": true,
            "counter": 0,
            "triggeredAt": 1653597636
          }
        }
      }
    },
    "metadata": {
      "reported": {
        "alarms": {
          "ALR008": {
            "alarmTriggered": {
              "timestamp": 1653598179
            },
            "counter": {
              "timestamp": 1653598179
            },
            "triggeredAt": {
              "timestamp": 1653598179
            }
          }
        }
      }
    },
    "version": 541
  },
  "timestamp": 1653598179,
  "clientToken": "a6r0h-f27b416a"
}
```

**Comment by jmklix** [comment]
P65201887

**Comment by jmklix** [comment]
What region is this happening to you in?

**Comment by psiphi75** [comment]
Our dev/test stack is in ap-southeast-2 and production in us-east-1.  Occurs for all of them.

**Comment by jmklix** [comment]
Sorry to keep asking you  for info, but can you get the trace-id, region and timestamp from cloudwatch from when this happened to you? 

If you haven't already you can [enable cloudwatch](https://docs.aws.amazon.com/iot/latest/developerguide/configure-logging.html) for IoT and then view the logs following these [steps](https://docs.aws.amazon.com/iot/latest/developerguide/cloud-watch-logs.html)

**Comment by psiphi75** [comment]
Nope, this should not be closed.  I gave my update 2 days prior to the bot.

By the way, I have tried reproducing a couple of times and I'm unable to reproduce consistently.  I'm continuing to work on reproducing it with additional logs.

**Comment by psiphi75** [comment]
@jmklix, here are the details you requested.  I will raise an issue internally with AWS support to provide further logs and details.

region: `ap-southeast-2`

One of the duplicates begin with the trace ID of `742e41cd-b2fe-cbb1-f4ba-e77023b6a9ae` (2022-06-20 20:51:26.268), this is when the device attempts to reconnect but fails.  The actual re-connection has the trace id of `665bdba0-c501-8d35-6325-a5ff0da142c5` (2022-06-20 20:51:26.381).

**Comment by TingDaoK** [comment]
> When the network is lost our application will explicitly call MqttConnection->Disconnect() and Connect() for when the network is available again.

The SDK will handle the reconnect logic. If you call MqttConnection->Disconnect() and Connect(), you will lose all the state. Don't call MqttConnection->Disconnect() and Connect() if you want to reconnect.

**Comment by jmklix** [comment]
As TingDaoK said you don't need (shouldn't be) calling the disconnect yourself as it is handled by the SDK. This did uncover a small cleanup problem which was fixed by this [PR](https://github.com/awslabs/aws-c-mqtt/pull/216/files). This is now in the latest release so even if you call disconnect you shouldn't see multiple shadow updates.
